### DISCUSSION TITLE
Declarative and imperative migrations

### DISCUSSION THREAD

**Discussion by terion-name** [directus/directus#11249] [Feature Requests]
Currently to transfer datamodel changes directus uses schema dump/apply. 

E.g. declarative migrations.

While declarative migrations are great for start of blank project, they absolutely suck at support/update cycle. Seriously, it is just impossible to make purely declarative migrations perfect. I've seen many attempts, I've used many variants, I've tried to do this myself as many of us — it just don't work. They are too rigid and can't handle a lot of cases. The most simple and obvious — renaming fields. For a declarative migrator it is deleting one field and creating other. And bye-bye data. And keys handling is terrible with declaration — planner simply can't figure out proper operation order to properly schedule foreign keys creation in interlinkied models (that can be even recursive). There are lots of cases when declarative don't work and attempts to fix them always lead to terrible hacks.

On the other hand we are too lazy to write imperative migrations — the whole point of directus is click-click and deploy.

The way to make everything work is to mix these approaches.

1. You have declarative migrations that can be like modules for bootstrapping new projects — great thing really, entire apps can be distributed as schema files (but they must have role and access blocks too)
2. There should be a command that converts current schema to set of knex migrations
3. There should be a command that renders changes as a schema patch and set of migrations

And then on deploy one just runs these migrations. As far as I can see (sorry, didn't investigate source in this place — overloaded with work) having schema parser and differ already this should not be a huge problem — you already run sql (probably via knex), just don't run it but write to files. 

And this gives best of all worlds:
1. You still have gui to build data in no time
2. You still have great declarative schemas to kickstart a project
3. BUT you have flexibility of imperative migrations — you simply can fix/change generated migrations by hand. If planner generated some incorrect actions (as it does often now) — you can clean migration file. For renaming actions you can remove drop/create actions and add alters. Also your migration flow becomes predictable and obvious, not a blackbox like now

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
